File: packages/typescript/src/webhook-signature.ts
/**
 * Webhook signature verification, usable without an API key.
 *
 * The API sends `X-OpenSMS-Signature: t=<unix seconds>,v1=<hex>` where
 * `v1 = hex(HMAC-SHA256(secret, "<t>.<raw body>"))` and `secret` is the full
 * `whsec_...` endpoint secret used verbatim. Parsing mirrors the server's
 * `internal/webhooks/signature.go`.
 * @module
 */
import { OpensmsError } from './errors.js';
import { fromWire, type WebhookEvent } from './models.js';
import { hmacSha256, toHex } from './internal/sha256.js';

/** Name of the signature header. */
export const SIGNATURE_HEADER = 'X-OpenSMS-Signature';

export interface VerifyOptions {
  /** Maximum age (and clock skew) in seconds. Defaults to `300`; the boundary is inclusive. */
  toleranceSeconds?: number;
  /** Current time in unix seconds. Defaults to the system clock. */
  now?: number;
}

type Outcome = 'valid' | 'invalid' | 'expired';

const encoder = new TextEncoder();
const decoder = new TextDecoder();

function bytesOf(payload: string | Uint8Array): Uint8Array {
  return typeof payload === 'string' ? encoder.encode(payload) : payload;
}

function parseHeader(header: string): { t: string; v1: string } | null {
  const fields: Record<string, string> = {};
  for (const rawPart of header.split(',')) {
    const part = rawPart.trim();
    const eq = part.indexOf('=');
    if (eq < 0) return null;
    const k = part.slice(0, eq);
    const v = part.slice(eq + 1);
    if (k === '' || v === '' || Object.prototype.hasOwnProperty.call(fields, k)) return null;
    fields[k] = v;
  }
  const keys = Object.keys(fields);
  if (keys.length !== 2 || fields.t === undefined || fields.v1 === undefined) return null;
  return { t: fields.t, v1: fields.v1 };
}

function constantTimeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  return diff === 0;
}

function check(
  payload: string | Uint8Array,
  header: string | null | undefined,
  secret: string,
  opts: VerifyOptions,
): Outcome {
  const tolerance = opts.toleranceSeconds ?? 300;
  if (typeof secret !== 'string' || secret.trim() === '' || !(tolerance >= 0)) return 'invalid';
  if (typeof header !== 'string' || header === '') return 'invalid';
  const parsed = parseHeader(header);
  if (!parsed) return 'invalid';
  // Same order as the server: timestamp, then freshness, then the digest.
  if (!/^[+-]?\d+$/.test(parsed.t)) return 'invalid';
  const t = Number(parsed.t);
  if (!Number.isSafeInteger(t)) return 'invalid';
  const now = opts.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - t) > tolerance) return 'expired';
  if (!/^[0-9a-fA-F]{64}$/.test(parsed.v1)) return 'invalid';
  const body = bytesOf(payload);
  const prefix = encoder.encode(`${parsed.t}.`);
  const message = new Uint8Array(prefix.length + body.length);
  message.set(prefix);
  message.set(body, prefix.length);
  const expected = toHex(hmacSha256(encoder.encode(secret), message));
  return constantTimeEqual(expected, parsed.v1.toLowerCase()) ? 'valid' : 'invalid';
}

/**
 * Return `true` when `header` is a valid, fresh signature of `payload`.
 * Pass the raw request body exactly as received, before any JSON parsing.
 */
export function verifySignature(
  payload: string | Uint8Array,
  header: string | null | undefined,
  secret: string,
  options: VerifyOptions = {},
): boolean {
  return check(payload, header, secret, options) === 'valid';
}

/**
 * Verify the signature and parse the event envelope. Throws
 * {@link OpensmsError} with `status 0` and `code` `invalid_signature` or
 * `expired_signature`. The envelope keys are camelCased; `data` is returned
 * exactly as sent.
 */
export function constructEvent(
  payload: string | Uint8Array,
  header: string | null | undefined,
  secret: string,
  options: VerifyOptions = {},
): WebhookEvent {
  const outcome = check(payload, header, secret, options);
  if (outcome === 'expired') {
    throw new OpensmsError({
      status: 0,
      code: 'expired_signature',
      message: 'Webhook signature timestamp is outside the tolerance window.',
    });
  }
  if (outcome !== 'valid') {
    throw new OpensmsError({ status: 0, code: 'invalid_signature', message: 'Webhook signature is invalid.' });
  }
  const text = typeof payload === 'string' ? payload : decoder.decode(payload);
  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch (err) {
    throw new OpensmsError({ status: 0, code: 'invalid_payload', message: 'Webhook body is not JSON.', cause: err });
  }
  const envelope = (raw ?? {}) as Record<string, unknown>;
  const { data, ...rest } = envelope;
  return { ...fromWire<Omit<WebhookEvent, 'data'>>(rest), data: (data ?? {}) as WebhookEvent['data'] };
}

File: packages/go/webhook_signature.go
package opensms

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"strconv"
	"strings"
	"time"
)

// SignatureHeader is the header that carries a webhook delivery signature.
const SignatureHeader = "X-OpenSMS-Signature"

// DefaultSignatureTolerance is the default allowed clock skew.
const DefaultSignatureTolerance = 300 * time.Second

// VerifyOptions tunes signature verification.
type VerifyOptions struct {
	// Tolerance is the allowed difference between now and the signed
	// timestamp (default 300s). The boundary is inclusive.
	Tolerance time.Duration
	// Now overrides the current time (for tests).
	Now time.Time
}

var (
	errInvalidSignature = errors.New("invalid signature")
	errExpiredSignature = errors.New("expired signature")
)

// VerifySignature reports whether header (the X-OpenSMS-Signature value) is
// a valid signature of payload, the exact raw request body, for the endpoint
// secret. The secret is used verbatim, including its whsec_ prefix. It needs
// no API key or client.
func VerifySignature(payload []byte, header, secret string, opts ...VerifyOptions) bool {
	return verifySignature(payload, header, secret, opts...) == nil
}

// ConstructEvent verifies the signature and then decodes the event envelope.
// A bad signature returns *Error with Status 0 and Code "invalid_signature";
// a timestamp outside the tolerance returns Code "expired_signature".
func ConstructEvent(payload []byte, header, secret string, opts ...VerifyOptions) (*WebhookEvent, error) {
	switch err := verifySignature(payload, header, secret, opts...); err {
	case nil:
	case errExpiredSignature:
		return nil, newLocalError("expired_signature", "webhook signature timestamp is outside the tolerance", err)
	default:
		return nil, newLocalError("invalid_signature", "webhook signature is invalid", err)
	}
	var ev WebhookEvent
	if err := json.Unmarshal(payload, &ev); err != nil {
		return nil, newLocalError("invalid_payload", "webhook payload is not valid JSON: "+err.Error(), err)
	}
	return &ev, nil
}

// verifySignature mirrors the server's VerifyWithTolerance exactly.
func verifySignature(payload []byte, header, secret string, opts ...VerifyOptions) error {
	tolerance := DefaultSignatureTolerance
	now := time.Now()
	if len(opts) > 0 {
		if opts[0].Tolerance > 0 {
			tolerance = opts[0].Tolerance
		}
		if !opts[0].Now.IsZero() {
			now = opts[0].Now
		}
	}
	if strings.TrimSpace(secret) == "" {
		return errInvalidSignature
	}
	values, ok := parseSignatureHeader(header)
	if !ok {
		return errInvalidSignature
	}
	ts, err := strconv.ParseInt(values["t"], 10, 64)
	if err != nil {
		return errInvalidSignature
	}
	when := time.Unix(ts, 0)
	if now.Sub(when) > tolerance || when.Sub(now) > tolerance {
		return errExpiredSignature
	}
	want, err := hex.DecodeString(values["v1"])
	if err != nil || len(want) != sha256.Size {
		return errInvalidSignature
	}
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(values["t"]))
	h.Write([]byte("."))
	h.Write(payload)
	if subtle.ConstantTimeCompare(h.Sum(nil), want) != 1 {
		return errInvalidSignature
	}
	return nil
}

// parseSignatureHeader requires exactly the keys t and v1, no empty keys or
// values and no duplicates.
func parseSignatureHeader(header string) (map[string]string, bool) {
	values := make(map[string]string, 2)
	for _, part := range strings.Split(header, ",") {
		pair := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(pair) != 2 || pair[0] == "" || pair[1] == "" {
			return nil, false
		}
		if _, dup := values[pair[0]]; dup {
			return nil, false
		}
		values[pair[0]] = pair[1]
	}
	if len(values) != 2 || values["t"] == "" || values["v1"] == "" {
		return nil, false
	}
	return values, true
}
